### PR TITLE
Add collaborator support

### PR DESCRIPTION
**Pull request by YLivay**
oh yes

### CODE REVIEW

**Comment by niksudan** [comment]
Could it be? Has the prophecy finally come true?!

File: src/app/views/dashboard/games/manage/game/packages/list/list.html [hunk 1]
```diff
@@ -24,34 +24,38 @@
...
-				class="btn btn-success-outline btn-block"
-				:to="{
-					name: 'dash.games.manage.game.packages.add',
-				}"
-				>
-				<app-jolticon icon="add" />
-				<translate>dash.games.packages.add_package_button</translate>
-			</router-link>
-		</div>
+		<app-game-perms required="all">
```
Comment (hworld) [inline src/app/views/dashboard/games/manage/game/packages/list/list.html]: tag="div" and get rid of the extra div?

File: src/app/views/dashboard/games/manage/game/packages/list/list.html [hunk 2]
```diff
@@ -24,34 +24,38 @@
...
+					<app-jolticon icon="add" />
+					<translate>dash.games.packages.add_package_button</translate>
+				</router-link>
+			</div>
+		</app-game-perms>
 		<br>
 
 		<app-card-list v-if="packages.length" :items="packages">
-			<app-card-list-draggable @change="saveSort">
+			<app-card-list-draggable :disabled="!game.hasPerms('details')" @change="saveSort">
```
Comment (hworld) [inline src/app/views/dashboard/games/manage/game/packages/list/list.html]: cachy cache

File: src/app/views/dashboard/games/manage/game/packages/list/list.html [hunk 3]
```diff
@@ -134,26 +138,30 @@
...
-							class="btn btn-outline"
-							:to="{
-								name: 'dash.analytics',
-								params: { resource: 'Game_Package', resourceId: package.id },
-							}"
-							>
-							<app-jolticon icon="chart" />
-							<translate>Analytics</translate>
-						</router-link>
+						<app-game-perms required="all">
```
Comment (hworld) [inline src/app/views/dashboard/games/manage/game/packages/list/list.html]: we could probably let anyone see the widget. you can't modify anything on that page

File: src/app/views/dashboard/games/manage/game/packages/list/list.ts [hunk 4]
```diff
@@ -57,6 +59,7 @@ export default class RouteDashGamesManageGamePackagesList extends BaseRouteCompo
 		const payload = await Api.sendRequest('/web/dash/developer/games/packages/' + route.params.id);
 
 		if (payload.packages && !payload.packages.length) {
+			// TODO(collaborators) a page for when there are no packages but no permission to create any
```
Comment (hworld) [inline src/app/views/dashboard/games/manage/game/packages/list/list.ts]: what do you think we should do? I don't know if we should have a whole new page for this, or just have the list page show "No packages" and don't do the redirect here if they don't have perms

File: src/app/views/dashboard/games/manage/manage.html [hunk 5]
```diff
@@ -96,28 +101,50 @@ <h1 class="section-header">
...
-				<li v-app-tooltip.bottom="$gettext( `Manage your game keys and give access to users.` )">
-					<router-link
-						:to="{ name: 'dash.games.manage.key-groups.list' }"
-						:class="{ active: $route.name.indexOf( 'dash.games.manage.key-groups' ) === 0 }"
-						>
-						<translate>Keys/Access</translate>
-					</router-link>
-				</li>
-				<li v-app-tooltip.bottom="$gettext( `Game Jolt Sites are customizable external sites for your portfolio and games!` )">
+				<app-game-perms required="game-api">
```
Comment (hworld) [inline src/app/views/dashboard/games/manage/manage.html]: this needs to be tag="li". you can't have any other tags besides <li> within <ul> according to the standard

File: src/app/views/discover/games/view/_nav/nav.html [hunk 6]
```diff
@@ -76,8 +76,20 @@
...
+						params: { id: game.id },
+					}"
+					>
+					<app-jolticon icon="cog" />
+					<translate>Manage Game</translate>
+				</router-link>
+			</app-game-perms>
 			<a class="list-group-item has-icon"
-				v-if="app.user && app.user.id !== game.developer.id"
+				v-if="app.user && !game.hasPerms()"
```
Comment (hworld) [inline src/app/views/discover/games/view/_nav/nav.html]: cache

File: src/app/views/discover/games/view/_nav/nav.html [hunk 7]
```diff
@@ -66,7 +66,7 @@
 				</router-link>
 			</li>
 
-			<li v-if="app.user && app.user.id !== game.developer.id">
+			<li>
```
Comment (hworld) [inline src/app/views/discover/games/view/_nav/nav.html]: we don't always have more options to show do we?

File: src/app/views/discover/games/view/download/build/build.html [hunk 8]
```diff
@@ -67,7 +67,7 @@ <h2 class="section-header">
 								<!--
 									Rating widget
 								-->
-								<template v-if="app.user && app.user.id !== game.developer.id && game.ratings_enabled && game.can_user_rate">
+								<template v-if="app.user && !game.hasPerms() && game.ratings_enabled && game.can_user_rate">
```
Comment (hworld) [inline src/app/views/discover/games/view/download/build/build.html]: we need a discussion about ratings as collaborators

File: src/app/views/dashboard/main/overview/overview.html [hunk 9]
```diff
@@ -393,6 +393,80 @@ <h2 :class="{ 'section-header': Screen.isDesktop }">
 					</div>
 				</template>
 
+				<!--
+					Manage Collaborations
+				-->
```
Comment (hworld) [inline src/app/views/dashboard/main/overview/overview.html]: merge this into the normal "Your Games" list and instead just show a tag underneath that says "Collaborator".

Allow them to stop collaborating when they're managing the game (in the game overview), not from this screen.

File: src/app/views/dashboard/games/manage/game/overview/overview.html [hunk 10]
```diff
@@ -150,92 +154,96 @@
...
-			<app-expand :when="!game.canceled">
 				<h4>
-					<translate>Cancel Game</translate>
+					<translate>Remove Game</translate>
 				</h4>
 
 				<div class="page-help">
-					<p><translate>Canceling your game will signal to everyone that you're no longer working on it. People will still be able to view the game page and access your published packages, but it will not show in game listings.</translate></p>
-					<p><translate>If you've transitioned development off of Game Jolt, you must remove your game instead.</translate></p>
+					<p v-translate>
```
Comment (hworld) [inline src/app/views/dashboard/games/manage/game/overview/overview.html]: Only the main game owner should be able to remove.

File: src/app/components/activity/feed/controls/controls.ts [hunk 11]
```diff
@@ -92,7 +92,11 @@ export class AppActivityFeedControls extends Vue {
 
 	get shouldShowStats() {
 		return (
-			!!this.post && this.showExtraInfo && !!this.app.user && this.post.user.id === this.app.user.id
+			!!this.post &&
+			this.showExtraInfo &&
+			!!this.app.user &&
+			// TODO(collaborators) can we shorten this to just this.game.hasPerms or this.post.game.hasPerms?
```
[the remark this answers is not in the document]
Comment (YLivay) [reply src/app/components/activity/feed/controls/controls.ts]: can we, forgot to check - will check. hi.

File: src/app/components/forms/game/game.html [hunk 12]
```diff
@@ -52,6 +52,7 @@
 					:rules="{
 						max: 250,
 					}"
+					:disabled="!model.hasPerms('details')"
```
[the remark this answers is not in the document]
Comment (YLivay) [reply src/app/components/forms/game/game.html]: We'd need getters for every unique type of check. is that sane?

File: src/app/components/forms/game/package/package.html [hunk 13]
```diff
@@ -19,6 +19,7 @@
 						initVal: method === 'edit' ? (model.title || game.title) : undefined,
 					},
 				}"
+				:disabled="!game.hasPerms('details')"
```
[the remark this answers is not in the document]
Comment (YLivay) [reply src/app/components/forms/game/package/package.html]: mkay so any package modifications is details, and then I can completely block community managers from going into the packages section?

File: src/app/components/game/perms/perms.ts [hunk 14]
```diff
@@ -0,0 +1,40 @@
...
+	@Prop({ type: String, default: 'span' })
+	tag: string;
+	@Prop(Boolean) debug?: boolean;
+
+	get targetGame() {
+		return this.game || this.storeGame;
+	}
+
+	get hasPerms() {
+		const perms: Perm[] = (this.required as any).split(',');
```
[the remark this answers is not in the document]
Comment (YLivay) [reply src/app/components/game/perms/perms.ts]: still need any to coerce it into perms. don't think we need to actually check it in this func because we'd be testing it already as we develop.

File: src/app/views/dashboard/games/manage/game/packages/list/list.ts (hunk 4 above)
Comment (YLivay) [reply src/app/views/dashboard/games/manage/game/packages/list/list.ts]: did this, didn't test. can you give it a check?
